### DISCUSSION TITLE
Why Airflow 2.x opens too many postgres db connection and throws PID of job runner does not match error?

### DISCUSSION THREAD

**Discussion by MhmtErsy** [apache/airflow#18320] [Q&A]
Hi all, 

As you know Airflow opens some connections to the metadata database. I am currently using 1.10.6 version on production and it opens about 5.000 connections during a DagRun that contains parallel 9 jobs (within a 9 slot pool). 
I have also installed a fresh Airflow 2.1.2 version on my development environment and I copied that Dag into there. Then the same way I started the DAG within 9 slots pool. But I noticed that it opens 40.000-45.000 sockets to reach my external Postgresql metadata database and some of the tasks got failed status without start and logging. I also saw "PID of job runner does not match error" error on celery worker log during this run.

What is the main reason behind Airflow is opening so many connections for the same number of parallel jobs in 2.0? Why does it need it? As far as I remember one of the new features of Airflow 2.0 is **[DAGs in Airflow 2.0 are serialized in the database](https://www.astronomer.io/blog/introducing-airflow-2-0#apache-airflow-110-vs-apache-airflow-20)**. Could it add extra load on the metadata database? 

My below airflow 2.1.2 and 1.10.6 airflow.cfg configurations are the same.

```
[core]
sql_alchemy_pool_enabled = True
sql_alchemy_pool_size = 50
sql_alchemy_max_overflow = 100
sql_alchemy_pool_recycle = 1800
parallelism = 256
dag_concurrency = 128
max_active_runs_per_dag = 16
sql_alchemy_conn = # a postgresql db
```

```
[celery]
worker_concurrency = 128
broker_url = # a redis db
result_backend = # a postgresql db
```

Note: I use CeleryExecutor. I check the connection number with the "`netstat -ton | grep WAIT | wc -l`" command on the celery worker node.

**Comment by potiuk** [comment]
For Airflow, the recommended prectice is to use PGBouncer in front of Postgres DB. This will limit the connections. Airflow 2 is much faster and reaches much more parallelism when running, so it is expected that there will be more connections opened.

PGBouncer ,mitigates that.

The PID of job runner was indeed happening and it's been solved (there were few cases for it). Very soon 2.1.4 will be released (days) and it should address most of such issues.

**Reply by MhmtErsy** [reply]
Hi Jarek, thank you for responding. 
I see, I will investigate and try the pgBouncer, Thank you. 
By the way, Firstly I had installed 2.1.3 but then I had to revert to 2.1.2 because of  #10026. Will there be an update about #10026 in 2.1.4?

**Comment by potiuk** [comment]
Not 100% sure if that's the same issue but similar problems were fixed

**Comment by potiuk** [comment]
You can take a look at the changelog in main branch yourself.
